src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Memcached/EnyimMemcachedCoreWrapper.cs

```diff
@@ -0,0 +1,97 @@
+            {
+                transaction.AttachToAsync();
+            }
+
+            // Internally, the key is used to determine what server to read from in a multi-server environment.
+            // Without a key, we can't determine the server, so we can't determine the connection info.
+
+            ParsedSqlStatement parsedStatement;
+            string key;
+
+            // Operation is the first argument in all cases, Key is the second argument
+            if (instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformStore")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformStoreAsync")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformMutate")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformMutateAsync")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformConcatenate"))
+            {
+                key = instrumentedMethodCall.MethodCall.MethodArguments[1].ToString();
+                parsedStatement = new ParsedSqlStatement(DatastoreVendor.Memcached,
+                    key,
+                    instrumentedMethodCall.MethodCall.MethodArguments[0].ToString());
+            }
+            // Operation is always Get, Key is the first argument
+            else if (instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformTryGet")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("PerformGet")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("GetAsync"))
+            {
+                key = instrumentedMethodCall.MethodCall.MethodArguments[0].ToString();
+                parsedStatement = new ParsedSqlStatement(DatastoreVendor.Memcached,
+                    key,
+                    "Get");
+            }
+            // Operation is always Remove, Key is the first argument
+            else if (instrumentedMethodCall.MethodCall.Method.MethodName.Equals("Remove")
+                || instrumentedMethodCall.MethodCall.Method.MethodName.Equals("RemoveAsync"))
+            {
+                key = instrumentedMethodCall.MethodCall.MethodArguments[0].ToString();
+                parsedStatement = new ParsedSqlStatement(DatastoreVendor.Memcached,
+                    key,
+                    "Remove");
+            }
+            // Should not happen
+            else
+            {
+                return Delegates.NoOp;
+            }
+
+            var connectionInfo = MemcachedHelpers.GetConnectionInfo(
+                key,
+                instrumentedMethodCall.MethodCall.InvocationTarget,
+                agent);
+
+            var segment = transaction.StartDatastoreSegment(instrumentedMethodCall.MethodCall, parsedStatement, connectionInfo, isLeaf: true);
+            segment.AddCustomAttribute("key", key); // Storing the key is optional, but could be useful and we already have it.
```

Comment:
Custom attributes are typically reserved for customer provided attributes, and there is a limit to the number of custom attributes that can be collected. Should this be collected as an agent attribute instead? What are other agents doing?

Reply:
Java does not store the key.  Node does capture it, but I am not able to determine what type of attribute they are storing it as.  Java also has different metrics compared to what I have setup, at least for statements.  I am going to look things over again to see if I might have missed something.